src/wasm-linker.cpp

```diff
@@ -49,13 +49,38 @@ void Linker::placeStackPointer(size_t stackAllocation) {
 }
 
 void Linker::layout() {
+  // Convert calls to undefined functions to call_imports
+  for (const auto& f : out.undefinedFunctions) {
+    Name target = f.first;
+    // Create an import for the target if necessary.
+    if (!out.wasm.checkImport(target)) {
+      auto import = out.wasm.allocator.alloc<Import>();
+      import->name = import->base = target;
+      import->module = ENV;
+      import->type = ensureFunctionType(getSig(*f.second.begin()), &out.wasm,
+                                        out.wasm.allocator);
+      out.wasm.addImport(import);
+    }
+    // Change each call. The target is the same since it's still the name.
+    // Delete and re-allocate the Expression as CallImport to avoid undefined
+    // behavior.
+    for (auto* call : f.second) {
+      Call callCopy = std::move(*call);
+      call->~Call();
+      CallImport* newCall = new (call) CallImport;
```

Comment:
`nop()` is totally undefined behavior because if you allocate an object with a type, you can't reference it through a pointer of another type (even if the classes have the same initial member declarations).

That's similar to one idea I originally had which was to just change the `_id` field from `CallId` to `CallImportId` since those classes have exactly the same member declarations. But in either case when you traverse the AST you end up calling `visitCallImport(CallImport* curr)` on an object allocated as a `Call` and you're in UB-land.

There is a provision in the Standard that says if you have a union of types that have a common initial sequence (of member declaration), then you can access those members that way. So we could make the visitor operate on such a union type (and extract pointers to the subtypes and pass them to the visitor methods), but it would require the allocations to all be of the union type (i.e. the size of the largest type, which I believe is a vector, a Name, and 2 pointers, plus the base `Expression`), which would cost more memory.

Another option would be to have parent/user pointers in the AST so that we could easily reallocate them and fix the parents (or have such pointers just for the calls). I would actually be OK with that; the single-use nature of an expression tree makes this rather nice because when you do want to replace a node you only have one user to fix up. But IIRC you've expressed a desire not to have parent pointers in the past.

Another option would be to do a full-blown traversal pass just to change these nodes; I would rather not have to traverse the whole AST, when I could just do what I did here; i.e. just keep a list of the nodes that need to be fixed.

Another option would be to do what I did here; a little bit of ugly reallocation but safe and localized.